False positive on __required__ and __optional__ on TypedDict.

### Bug description

```python
from typing import TypedDict, Required, NotRequired

class MyDict(TypedDict):
    required: Required[str]
    not_required: NotRequired[str]
    implicitely_required: str

def main():
    print(MyDict.__required_keys__)
    print(MyDict.__optional_keys__)

if __name__ == "__main__":
    main()
```

### Configuration

```ini

```

### Command used

```shell
pylint --disable=missing-docstring main.py
```

### Pylint output

```python
************* Module main
main.py:9:10: E1101: Class 'MyDict' has no '__required_keys__' member (no-member)
main.py:10:10: E1101: Class 'MyDict' has no '__optional_keys__' member (no-member)
```

### Expected behavior

Pylint shouldn't report it as a missing member as the code runs without any issue.

It seems the same issue was already fixed for[ __annotations__ ](https://github.com/pylint-dev/pylint/issues/7126)



### Pylint version

```shell
pylint 3.3.3
astroid 3.3.8
Python 3.11.10 (main, Sep  7 2024, 01:03:31) [Clang 15.0.0 (clang-1500.3.9.4)]
```

### OS / Environment

Darwin 24.2.0

### Additional dependencies

```python

```

--- a/ChangeLog
+++ b/ChangeLog
@@ -7,6 +7,13 @@ What's New in astroid 4.2.0?
 ============================
 Release date: TBA
 
+
+* Add ``__required_keys__`` and ``__optional_keys__`` to the inferred
+  ``TypedDict`` base class, so subclasses no longer raise a ``no-member``
+  false positive in pylint when accessing those attributes at runtime.
+
+  Closes pylint-dev/pylint#10158
+
 * Comprehension conditions now constrain the inference of names used in the
   comprehension, like ``if`` statement conditions already did:
   ``[x for x in lst if x is not None]`` no longer infers ``None`` for ``x``

--- a/astroid/brain/brain_typing.py
+++ b/astroid/brain/brain_typing.py
@@ -229,6 +229,12 @@ def infer_typedDict(  # pylint: disable=invalid-name
     class_def.postinit(bases=[extract_node("dict")], body=[], decorators=None)
     func_to_add = _extract_single_node("dict")
     class_def.locals["__call__"] = [func_to_add]
+    # TypedDict subclasses have ``__required_keys__`` and ``__optional_keys__``
+    # class attributes at runtime (e.g. ``MyDict.__required_keys__``), even
+    # though the annotation-only body never declares them.
+    for attr in ("__required_keys__", "__optional_keys__"):
+        func_to_add = _extract_single_node("dict")
+        class_def.locals[attr] = [func_to_add]
     return iter([class_def])
 
 

--- a/tests/brain/test_brain.py
+++ b/tests/brain/test_brain.py
@@ -655,6 +655,22 @@ class TypingBrain(unittest.TestCase):
         # Test TypedDict instance is callable
         assert next(code[1].infer()).callable() is True
 
+    def test_typed_dict_required_and_optional_keys(self):
+        """TypedDict subclasses expose ``__required_keys__`` and ``__optional_keys__``."""
+        code = builder.extract_node("""
+        from typing import TypedDict
+        from typing_extensions import TypedDict as ETD
+
+        class CustomTD(TypedDict):  #@
+            var: int
+
+        class CustomTD2(ETD):  #@
+            var: int
+        """)
+        for cls in code:
+            for attr in ("__required_keys__", "__optional_keys__"):
+                assert cls.getattr(attr), f"{cls.name}.{attr} not found"
+
     def test_typing_alias_type(self):
         """
         Test that the type aliased thanks to typing._alias function are

